net/FlatBuffers/ByteBuffer.cs

```diff
@@ -42,20 +42,28 @@
 using System.Runtime.InteropServices;
 using System.Text;
 
+#if ENABLE_SPAN_T
+using System.Buffers.Binary;
+#endif
+
 #if ENABLE_SPAN_T && !UNSAFE_BYTEBUFFER
 #error ENABLE_SPAN_T requires UNSAFE_BYTEBUFFER to also be defined
 #endif
 
 namespace FlatBuffers
 {
-    public abstract class ByteBufferAllocator : IDisposable
+    public abstract class ByteBufferAllocator
     {
-#if UNSAFE_BYTEBUFFER
-        public unsafe byte* Buffer
-        {
-            get;
-            protected set;
-        }
+#if ENABLE_SPAN_T
+        public abstract Span<byte> GetSpan(int start, int length);
+        public abstract ReadOnlySpan<byte> GetReadOnlySpan(int start, int length);
```

Comment:
Do we need both the read-only and writable variants?  There shouldn't be a penalty associated with the implicit cast from Span to ReadOnlySpan.

Reply:
Yes, if we want to enable custom `ByteBufferAllocator` implementations to have read-only stores. (It is slightly a misnomer to call it an "allocator". In reality, I think it would make sense to call it a "store" or "storage".)

Read-only stores is a scenario for the [Apache Arrow zero-copy reads](https://issues.apache.org/jira/browse/ARROW-4502) task I'm working on.

In that case, when reading an Arrow stream, we don't need to write to the `ByteBuffer`, we only need to read. So calling `GetSpan` is going to throw a `NotSupportedException`. This enables the Arrow stream reader to take in a `ReadOnlyMemory<byte>` object and read a flatbuffer `ByteBuffer` over top of it.

If we only had a writable `GetSpan` API - that scenario wouldn't be possible. I don't think you can get a writable `Span<byte>` from a `ReadOnlyMemory<byte>`.

Reply:
> I don't think you can get a writable Span<byte> from a ReadOnlyMemory<byte>.

You can, but it's unsafe and obviously dangerous.

> So calling GetSpan is going to throw a NotSupportedException.

Shouldn't there be some property then that states whether you can use the writable members? e.g. CanWrite? CanGetWritableMemory? etc.?

Reply:
I would think the same code that is creating the ByteBuffer is the one that is using it. If they create a read-only ByteBuffer, they probably shouldn't be using it to write.

This is sort of the same [discussion about removing `IDisposable`](https://github.com/google/flatbuffers/pull/5191#discussion_r258586469). If you created a disposable allocator, you should dispose of it yourself.

I think we can add a `CanWrite` API in the future, if it is necessary.

Reply:
> I would think the same code that is creating the ByteBuffer is the one that is using it.

If that's true, it'd be nice to avoid the ByteBuffer abstraction entirely :)

> I think we can add a CanWrite API in the future, if it is necessary.

Ok

Reply:
`ByteBuffer` was put in originally to make it easier to share the code generator between Java and C#, as they were deemed similar. Since then, the two have diverged more and more, and there was a plan at some point to fork the code generator for the 2 cases, to make it easier for C# to go its own path. That way "inlining" ByteBuffer would also become an option.

Reply:
`ByteBuffer` is a pretty core type in Flat Buffers, from what I can tell. You basically need it in order to do anything with the API.
For example, here's how Apache Arrow is reading an instance of its `Message` table:

https://github.com/apache/arrow/blob/9600d8b85d44af8d59f1f680341021717738a7b1/csharp/src/Apache.Arrow/Ipc/ArrowStreamReader.cs#L70

You need to pass a `ByteBuffer` into `Flatbuf.XXX.GetRootAsXXX(buffer)` method in order to read a table.

Reply:
Ah, hadn't realized we had exposed it. Still, some amount of inlining/refactoring is probably possible if needed, especially if we fork the generator.